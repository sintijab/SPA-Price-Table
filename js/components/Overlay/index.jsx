import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { Input, ButtonBase } from '@material-ui/core';
import DOMPurify from 'dompurify';
import { updateData } from '../../utils/functions';
import { validatePrice } from '../../utils/priceFunctions';
import { URL_HOST, ALL_COMPANIES, BUDGET } from '../../constants/endpoints';
import {
  CLOSE_BUTTON,
  INPUT_ERROR_TEXT,
  INPUT_VALID_TEXT,
} from '../../constants/titles';
import styles from './index.scss';

const Overlay = ({ item, hideModal, updateBudget }) => {
  const {
    register,
    errors,
    handleSubmit,
  } = useForm({
    mode: 'onChange',
    reValidateMode: 'onChange',
  });
  const [infoMessage, setAdditionalInfo] = useState(false);
  const {
    name,
    budget,
    id,
  } = item || {};
  const budgetTotal = validatePrice(budget);
  const onSubmit = (e) => {
    const params = {
      url: `${URL_HOST}${ALL_COMPANIES}/${id}`,
      property: BUDGET,
      value: e.totalBudget,
    };
    updateData(params)
      .then(() => {
        setAdditionalInfo(INPUT_VALID_TEXT);
        updateBudget(id, e.totalBudget);
      }).catch((err) => {
        console.error(err);
      });
  };
  const closeOverlay = () => {
    setAdditionalInfo(null);
    hideModal();
  };
  const buttonText = `${errors.totalBudget ? 'Disabled' : 'Submit'}`;
  const { sanitize } = DOMPurify;
  return (
    <div className={styles.modal}>
      <div className="close-btn" role="presentation" onClick={() => closeOverlay()}>
        {CLOSE_BUTTON}
      </div>
      <div
        className={styles.modal_item}
        // eslint-disable-next-line
        dangerouslySetInnerHTML={{ __html: sanitize(`<h3>${name}</h3>`) }}
      />
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          name="totalBudget"
          type="number"
          inputProps={{
            step: 'any',
          }}
          defaultValue={budgetTotal}
          onChange={() => setAdditionalInfo(null)}
          label="total price"
          inputRef={register({
            required: true,
            maxLength: 20,
            pattern: /^[0-9]+(\.[0-9][0-9])?$/,
          })}
          error={!!errors.totalBudget}
        />
        <br />
        {errors.totalBudget && (
          <h4 className="error-text">{INPUT_ERROR_TEXT}</h4>
        )}
        {!!infoMessage && (
          <h4 className="valid-text">{infoMessage}</h4>
        )}
        <ButtonBase
          type="submit"
          disabled={!!errors.totalBudget}
          classes={{ root: 'button' }}
        >
          {buttonText}
        </ButtonBase>
      </form>
    </div>
  );
};

Overlay.propTypes = {
  item: PropTypes.objectOf(PropTypes.oneOfType(
    [PropTypes.string, PropTypes.number],
  )),
  hideModal: PropTypes.func,
  updateBudget: PropTypes.func,
};

Overlay.defaultProps = {
  item: {},
  hideModal: () => {},
  updateBudget: () => {},
};

export default Overlay;
