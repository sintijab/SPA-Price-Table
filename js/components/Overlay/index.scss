@import '../../../styles/main';

.modal {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  margin: 1rem;
  padding: 10rem;
  border-radius: 1rem;
  line-height: 2;
  background: $white;
}

.modal_item {
  display: block;
}
