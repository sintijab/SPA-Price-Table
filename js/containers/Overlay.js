import { connect } from 'react-redux';
import { updateTotalBudget } from '../actions/priceActions';
import { closeModal } from '../actions/modalActions';
import Overlay from '../components/Overlay';


const mapDispatchToProps = (dispatch) => ({
  hideModal: () => dispatch(closeModal()),
  updateBudget: (id, price) => dispatch(updateTotalBudget(id, price)),
});

export default connect(
  null,
  mapDispatchToProps,
)(Overlay);
