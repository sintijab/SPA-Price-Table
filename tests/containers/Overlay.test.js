import React from 'react';
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import Provider from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';
import Overlay from '../../js/containers/Overlay';
import toJson from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureMockStore([thunk]);
const UserContext = React.createContext();

describe('<Overlay />', () => {
  let wrapper, store;

  beforeEach(() => {
    const initialState  = {
      modals: []
    }
    store = mockStore(initialState);
    wrapper = shallow(
      <Overlay store={store} />
    );
  });

    it('should render Overlay container', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
});
