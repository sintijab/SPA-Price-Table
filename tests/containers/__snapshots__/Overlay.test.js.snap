// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay /> should render Overlay container 1`] = `
<Overlay
  hideModal={[Function]}
  item={Object {}}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  updateBudget={[Function]}
/>
`;
